port/port_stdcxx.h

```diff
@@ -200,6 +203,113 @@ inline bool Zstd_Uncompress(const char* input, size_t length, char* output) {
 #endif  // HAVE_ZSTD
 }
 
+inline bool Zlib_Compress(int level, const char* input, size_t length,
+                          std::string* output, uint8_t raw = 0) {
+#if HAVE_ZLIB
+  const size_t BUFSIZE = 128 * 1024;
+  unsigned char temp_buffer[BUFSIZE];
+  // reserve enough memory to not reallocate on the fly
+  output->reserve(output->size() + compressBound(length));
+  z_stream strm;
+  strm.zalloc = 0;
+  strm.zfree = 0;
+  strm.next_in = (unsigned char*)(input);
+  strm.avail_in = (uint32_t)length;
+  strm.next_out = temp_buffer;
+  strm.avail_out = BUFSIZE;
+  auto res = deflateInit2(&strm, level, Z_DEFLATED, raw ? -15 : 15, 8,
+                          Z_DEFAULT_STRATEGY);
+  if (res != Z_OK) {
+    return false;
+  }
+  int deflate_res = Z_OK;
+  while (strm.avail_in != 0) {
+    int res = deflate(&strm, Z_NO_FLUSH);
+    if (res != Z_OK) {
+      return false;
+    }
+    if (strm.avail_out == 0) {
+      output->append(temp_buffer, temp_buffer + BUFSIZE);
+      strm.next_out = temp_buffer;
+      strm.avail_out = BUFSIZE;
+    }
+  }
+  while (deflate_res == Z_OK) {
+    if (strm.avail_out == 0) {
+      output->append(temp_buffer, temp_buffer + BUFSIZE);
+      strm.next_out = temp_buffer;
+      strm.avail_out = BUFSIZE;
+    }
+    deflate_res = deflate(&strm, Z_FINISH);
+  }
+  if (deflate_res != Z_STREAM_END) {
+    return false;
+  }
+  output->append(temp_buffer, temp_buffer + BUFSIZE - strm.avail_out);
+  deflateEnd(&strm);
+  return true;
+#else
+  // Silence compiler warnings about unused arguments.
+  (void)level;
+  (void)input;
+  (void)length;
+  (void)output;
+  (void)raw;
+  return false;
+#endif  // HAVE_ZLIB
+}
+
+inline bool Zlib_Uncompress(const char* input, size_t length,
+                            std::string* output, uint8_t raw = 0)
+{
+#if HAVE_ZLIB
+  const int CHUNK = 64 * 1024;
+  int ret;
+  size_t have;
+  z_stream strm;
+  unsigned char out[CHUNK];
+  /* allocate inflate state */
+  strm.zalloc = Z_NULL;
+  strm.zfree = Z_NULL;
+  strm.opaque = Z_NULL;
+  strm.avail_in = (uint32_t)length;
+  strm.next_in = (Bytef*)input;
+  ret = inflateInit2(&strm, (raw ? -15 : 15));
+  if (ret != Z_OK) {
+    return false;
+  }
+  /* decompress until deflate stream ends or end of file */
+  do {
+    /* run inflate() on input until output buffer not full */
+    do {
+      strm.avail_out = CHUNK;
+      strm.next_out = out;
+      ret = ::inflate(&strm, Z_NO_FLUSH);
+      if (ret == Z_NEED_DICT) {
+        ret = Z_DATA_ERROR;
+      }
+      if (ret < 0) {
+        (void)inflateEnd(&strm);
+        return false;
+      }
+      have = CHUNK - strm.avail_out;
+      output->append((char*)out, have);
+    } while (strm.avail_out == 0);
+    /* done when inflate() says it's done */
+  } while (ret != Z_STREAM_END);
+  /* clean up and return */
+  (void)inflateEnd(&strm);
+  return ret == Z_STREAM_END ? true : false;
```

Comment:
```suggestion
  return ret == Z_STREAM_END;
```

Reply:
Is there any case where it could be different than Z_STREAM_END? I don't see a break in the loop and it would only exit if ret is exactly that value.
```suggestion
  return true;
```

Reply:
Read through the doc and turns out the outer while loop isn't required it was there in case we were reading a file and there was more data to read into input. As a sanity check I left return as what @Erik-Schulz suggested.